### PR TITLE
[TensorExpr] Add a pass to symbolize an input dimension.

### PR DESCRIPTION
**Pull request by ZolotukhinM**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #66522
* #66847
* #66515
* #66514
* #69055
* __->__ #68857
* #68856
* #68855
* #68854

Differential Revision: [D32632908](https://our.internmc.facebook.com/intern/diff/D32632908)

### CODE REVIEW

**Comment by ZolotukhinM** [comment]
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32632908).

**Comment by ZolotukhinM** [comment]
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32632908).

**Comment by ZolotukhinM** [comment]
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32632908).

**Comment by ZolotukhinM** [comment]
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32632908).

File: torch/csrc/jit/tensorexpr/graph_opt.h [hunk 1]
```diff
@@ -64,6 +64,10 @@ TORCH_API void annotateInputShapes(
 TORCH_API std::shared_ptr<Graph> removeUnusedSelfArgument(
     const std::shared_ptr<Graph>& graph);
 
+TORCH_API std::vector<int64_t> makeShapesSymbolic(
```
Comment (navahgar) [inline torch/csrc/jit/tensorexpr/graph_opt.h:67]: Can you add a comment here describing what this function does?

File: torch/csrc/jit/tensorexpr/graph_opt.cpp [hunk 2]
```diff
@@ -206,6 +206,56 @@ std::shared_ptr<Graph> removeUnusedSelfArgument(
   return graph;
 }
 
+std::vector<int64_t> makeShapesSymbolic(
+    std::shared_ptr<Graph>& graph,
+    const std::vector<int64_t>& size_vals) {
+  std::unordered_set<Value*> values;
```
Comment (navahgar) [inline torch/csrc/jit/tensorexpr/graph_opt.cpp:212]: We should probably consider exposing `all_values` from `Graph`: https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/ir/ir.h#L1160

**Review by navahgar — APPROVED** [review]
LGTM

**Review by eellison — COMMENTED** [review]
do we still need to do this ?? this is pretty far on the not proving correctness end

**Comment by ZolotukhinM** [comment]
> do we still need to do this ?? this is pretty far on the not proving correctness end

I would strongly prefer to have this. I have just checked standard JIT passes (I kept this one but made it only affect graph inputs and then ran the standard shape-prop) on a newer version of the model and they still can not handle it. This is indeed not a generally correct pass and should be used with caution, and I will be happy to replace it with a proper solution once it's there. But I believe that it's relatively easy to use this workaround in a "safe" way.

File: torch/csrc/jit/tensorexpr/graph_opt.cpp (hunk 2 above)
Comment (ZolotukhinM) [reply torch/csrc/jit/tensorexpr/graph_opt.cpp:212]: I think it's a relatively rare case when we need to iterate over all values (and not nodes), so I'm not sure it's worth exposing (also, I think a common pattern "for each graph input, then for each input and output of each graph node, then for each graph output" works well for most cases and frankly could've been used here as well).

**Comment by eellison** [comment]
For the older model - this was the extent of the changes: https://github.com/pytorch/pytorch/pull/68826/files 

I think this amount of work should be a reasonable amount of work in relation to going to a solution that is not provably correct and doesnt extend to end users

**Comment by ZolotukhinM** [comment]
Yes, the amount of changes required to support the new version is small (I think we just need `quantized::linear` and `quantized::linear_relu`), no questions about that. The problem I'm trying to solve with this workaround is to get rid of shape-prop (and having to make JIT changes in general) from our critical path - having to commit something to pytorch every time a model is updated is very prohibitively slow for our use-case. As I mentioned earlier, having dynamic registration would have solved this particular case, but we don't have it yet.

In some sense it's similar to "torch.jit.trace" vs "torch.jit.script" - the latter is more "correct", however if we force all users to always use that, we can quickly lose our users.

**Comment by ZolotukhinM** [comment]
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32632908).

**Comment by eellison** [comment]
The dynamic registration is doable but a little tricky, because `linear` for instance calls into `matmul` which is pretty complicated and we only to define once. So the ideal registration would look something like:
```
quantized_linear(input, weight, bias):
    return torch.jit.shape_functions.linear(input, weight, bias)
```

It's doable but we need to decide on which primitives to expose (probably not all of them ?) and find a way for the different compilation units to play nicely together. See https://github.com/pytorch/pytorch/issues/62775.. 

For why i haven't done this yet.

**Comment by ZolotukhinM** [comment]
> For why i haven't done this yet.

Yeah, sorry if I appeared blaming - that was not my intention, and I understand technical aspects of why implementing it is not trivial. My point was that until we have a way for users to unblock themselves using a proper general solution, we should not shy away from providing partial solutions (as long as they do not move us in the opposite direction).

**Comment by ZolotukhinM** [comment]
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32632908).

**Comment by ZolotukhinM** [comment]
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32632908).

**Pull request closed without merging**